File: src/main/java/no/nav/bidrag/beregn/kostnadsberegnetbidrag/KostnadsberegnetBidragCoreImpl.java
package no.nav.bidrag.beregn.kostnadsberegnetbidrag;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import no.nav.bidrag.beregn.felles.FellesCore;
import no.nav.bidrag.beregn.felles.bo.Avvik;
import no.nav.bidrag.beregn.felles.bo.Periode;
import no.nav.bidrag.beregn.felles.dto.PeriodeCore;
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.bo.BPsAndelUnderholdskostnadPeriode;
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.bo.BeregnKostnadsberegnetBidragGrunnlag;
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.bo.BeregnetKostnadsberegnetBidragResultat;
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.bo.ResultatPeriode;
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.bo.SamvaersfradragPeriode;
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.bo.UnderholdskostnadPeriode;
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.dto.BPsAndelUnderholdskostnadPeriodeCore;
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.dto.BeregnKostnadsberegnetBidragGrunnlagCore;
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.dto.BeregnetKostnadsberegnetBidragResultatCore;
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.dto.ResultatBeregningCore;
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.dto.ResultatPeriodeCore;
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.dto.SamvaersfradragPeriodeCore;
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.dto.UnderholdskostnadPeriodeCore;
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.periode.KostnadsberegnetBidragPeriode;

public class KostnadsberegnetBidragCoreImpl extends FellesCore implements KostnadsberegnetBidragCore {

  private final KostnadsberegnetBidragPeriode kostnadsberegnetBidragPeriode;

  public KostnadsberegnetBidragCoreImpl(KostnadsberegnetBidragPeriode kostnadsberegnetBidragPeriode) {
    this.kostnadsberegnetBidragPeriode = kostnadsberegnetBidragPeriode;
  }

  public BeregnetKostnadsberegnetBidragResultatCore beregnKostnadsberegnetBidrag(
      BeregnKostnadsberegnetBidragGrunnlagCore beregnKostnadsberegnetBidragGrunnlagCore) {
    var beregnKostnadsberegnetBidragGrunnlag = mapTilBusinessObject(beregnKostnadsberegnetBidragGrunnlagCore);
    var beregnKostnadsberegnetBidragResultat = new BeregnetKostnadsberegnetBidragResultat(Collections.emptyList());
    var avvikListe = kostnadsberegnetBidragPeriode.validerInput(beregnKostnadsberegnetBidragGrunnlag);
    if (avvikListe.isEmpty()) {
      beregnKostnadsberegnetBidragResultat = kostnadsberegnetBidragPeriode.beregnPerioder(beregnKostnadsberegnetBidragGrunnlag);
    }
    return mapFraBusinessObject(avvikListe, beregnKostnadsberegnetBidragResultat);
  }

  private BeregnKostnadsberegnetBidragGrunnlag mapTilBusinessObject(BeregnKostnadsberegnetBidragGrunnlagCore beregnKostnadsberegnetBidragGrunnlagCore) {
    var beregnDatoFra = beregnKostnadsberegnetBidragGrunnlagCore.getBeregnDatoFra();
    var beregnDatoTil = beregnKostnadsberegnetBidragGrunnlagCore.getBeregnDatoTil();
    var soknadsbarnPersonId = beregnKostnadsberegnetBidragGrunnlagCore.getSoknadsbarnPersonId();
    var underholdskostnadPeriodeListe = mapUnderholdskostnadPeriodeListe(
        beregnKostnadsberegnetBidragGrunnlagCore.getUnderholdskostnadPeriodeListe());
    var bPsAndelUnderholdskostnadPeriodeListe = mapBPsAndelUnderholdskostnadPeriodeListe(
        beregnKostnadsberegnetBidragGrunnlagCore.getBPsAndelUnderholdskostnadPeriodeListe());
    var samvaersfradragPeriodeListe = mapSamvaersfradragPeriodeListe(
        beregnKostnadsberegnetBidragGrunnlagCore.getSamvaersfradragPeriodeListe());

    return new BeregnKostnadsberegnetBidragGrunnlag(beregnDatoFra, beregnDatoTil, soknadsbarnPersonId, underholdskostnadPeriodeListe,
        bPsAndelUnderholdskostnadPeriodeListe, samvaersfradragPeriodeListe);
  }

  private List<UnderholdskostnadPeriode> mapUnderholdskostnadPeriodeListe(List<UnderholdskostnadPeriodeCore> underholdskostnadPeriodeListeCore) {
    var underholdskostnadPeriodeListe = new ArrayList<UnderholdskostnadPeriode>();
    for (UnderholdskostnadPeriodeCore underholdskostnadPeriodeCore : underholdskostnadPeriodeListeCore) {
      underholdskostnadPeriodeListe.add(new UnderholdskostnadPeriode(
          underholdskostnadPeriodeCore.getReferanse(),
          new Periode(underholdskostnadPeriodeCore.getPeriode().getDatoFom(), underholdskostnadPeriodeCore.getPeriode().getDatoTil()),
          underholdskostnadPeriodeCore.getBelop()));
    }
    return underholdskostnadPeriodeListe;
  }

  private List<BPsAndelUnderholdskostnadPeriode> mapBPsAndelUnderholdskostnadPeriodeListe(
      List<BPsAndelUnderholdskostnadPeriodeCore> bPsAndelUnderholdskostnadPeriodeListeCore) {
    var bPsAndelUnderholdskostnadPeriodeListe = new ArrayList<BPsAndelUnderholdskostnadPeriode>();
    for (BPsAndelUnderholdskostnadPeriodeCore bPsAndelUnderholdskostnadPeriodeCore : bPsAndelUnderholdskostnadPeriodeListeCore) {
      bPsAndelUnderholdskostnadPeriodeListe.add(new BPsAndelUnderholdskostnadPeriode(
          bPsAndelUnderholdskostnadPeriodeCore.getReferanse(),
          new Periode(bPsAndelUnderholdskostnadPeriodeCore.getPeriode().getDatoFom(), bPsAndelUnderholdskostnadPeriodeCore.getPeriode().getDatoTil()),
          bPsAndelUnderholdskostnadPeriodeCore.getAndelProsent()));
    }
    return bPsAndelUnderholdskostnadPeriodeListe;
  }

  private List<SamvaersfradragPeriode> mapSamvaersfradragPeriodeListe(List<SamvaersfradragPeriodeCore> samvaersfradragPeriodeListeCore) {
    var samvaersfradragPeriodeListe = new ArrayList<SamvaersfradragPeriode>();
    for (SamvaersfradragPeriodeCore samvaersfradragPeriodeCore : samvaersfradragPeriodeListeCore) {
      samvaersfradragPeriodeListe.add(new SamvaersfradragPeriode(
          samvaersfradragPeriodeCore.getReferanse(),
          new Periode(samvaersfradragPeriodeCore.getPeriode().getDatoFom(), samvaersfradragPeriodeCore.getPeriode().getDatoTil()),
          samvaersfradragPeriodeCore.getBelop()));
    }
    return samvaersfradragPeriodeListe;
  }

  private BeregnetKostnadsberegnetBidragResultatCore mapFraBusinessObject(List<Avvik> avvikListe, BeregnetKostnadsberegnetBidragResultat resultat) {
    return new BeregnetKostnadsberegnetBidragResultatCore(mapResultatPeriode(resultat.getResultatPeriodeListe()), mapAvvik(avvikListe));
  }

  private List<ResultatPeriodeCore> mapResultatPeriode(List<ResultatPeriode> resultatPeriodeListe) {
    var resultatPeriodeCoreListe = new ArrayList<ResultatPeriodeCore>();
    for (ResultatPeriode resultatPeriode : resultatPeriodeListe) {
      var kostnadsberegnetBidragResultat = resultatPeriode.getResultat();
      resultatPeriodeCoreListe.add(new ResultatPeriodeCore(
          resultatPeriode.getSoknadsbarnPersonId(),
          new PeriodeCore(resultatPeriode.getPeriode().getDatoFom(), resultatPeriode.getPeriode().getDatoTil()),
          new ResultatBeregningCore(kostnadsberegnetBidragResultat.getBelop()),
          mapReferanseListe(resultatPeriode)));
    }
    return resultatPeriodeCoreListe;
  }

  private List<String> mapReferanseListe(ResultatPeriode resultatPeriode) {
    var resultatGrunnlag = resultatPeriode.getGrunnlag();

    var referanseListe = new ArrayList<String>();
    referanseListe.add(resultatGrunnlag.getUnderholdskostnad().getReferanse());
    referanseListe.add(resultatGrunnlag.getBPsAndelUnderholdskostnad().getReferanse());
    referanseListe.add(resultatGrunnlag.getSamvaersfradrag().getReferanse());
    return referanseListe;
  }
}

File: src/main/kotlin/no/nav/bidrag/beregn/kostnadsberegnetbidrag/KostnadsberegnetBidragCoreImpl.kt
package no.nav.bidrag.beregn.kostnadsberegnetbidrag

import no.nav.bidrag.beregn.felles.FellesCore
import no.nav.bidrag.beregn.felles.bo.Avvik
import no.nav.bidrag.beregn.felles.bo.Periode
import no.nav.bidrag.beregn.felles.dto.PeriodeCore
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.bo.BPsAndelUnderholdskostnadPeriode
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.bo.BeregnKostnadsberegnetBidragGrunnlag
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.bo.BeregnetKostnadsberegnetBidragResultat
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.bo.ResultatPeriode
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.bo.SamvaersfradragPeriode
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.bo.UnderholdskostnadPeriode
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.dto.BPsAndelUnderholdskostnadPeriodeCore
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.dto.BeregnKostnadsberegnetBidragGrunnlagCore
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.dto.BeregnetKostnadsberegnetBidragResultatCore
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.dto.ResultatBeregningCore
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.dto.ResultatPeriodeCore
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.dto.SamvaersfradragPeriodeCore
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.dto.UnderholdskostnadPeriodeCore
import no.nav.bidrag.beregn.kostnadsberegnetbidrag.periode.KostnadsberegnetBidragPeriode

class KostnadsberegnetBidragCoreImpl(private val kostnadsberegnetBidragPeriode: KostnadsberegnetBidragPeriode) : FellesCore(),
    KostnadsberegnetBidragCore {
    override fun beregnKostnadsberegnetBidrag(grunnlag: BeregnKostnadsberegnetBidragGrunnlagCore): BeregnetKostnadsberegnetBidragResultatCore {
        val beregnKostnadsberegnetBidragGrunnlag = mapTilBusinessObject(grunnlag)
        val avvikListe = kostnadsberegnetBidragPeriode.validerInput(beregnKostnadsberegnetBidragGrunnlag)
        val beregnKostnadsberegnetBidragResultat =
            if (avvikListe.isEmpty()) {
                kostnadsberegnetBidragPeriode.beregnPerioder(beregnKostnadsberegnetBidragGrunnlag)
            } else {
                BeregnetKostnadsberegnetBidragResultat(emptyList())
            }
        return mapFraBusinessObject(avvikListe, beregnKostnadsberegnetBidragResultat)
    }

    private fun mapTilBusinessObject(grunnlag: BeregnKostnadsberegnetBidragGrunnlagCore) =
        BeregnKostnadsberegnetBidragGrunnlag(
            beregnDatoFra = grunnlag.beregnDatoFra,
            beregnDatoTil = grunnlag.beregnDatoTil,
            soknadsbarnPersonId = grunnlag.soknadsbarnPersonId,
            underholdskostnadPeriodeListe = mapUnderholdskostnadPeriodeListe(grunnlag.underholdskostnadPeriodeListe),
            bPsAndelUnderholdskostnadPeriodeListe = mapBPsAndelUnderholdskostnadPeriodeListe(grunnlag.bPsAndelUnderholdskostnadPeriodeListe),
            samvaersfradragPeriodeListe = mapSamvaersfradragPeriodeListe(grunnlag.samvaersfradragPeriodeListe)
        )

    private fun mapFraBusinessObject(avvikListe: List<Avvik>, resultat: BeregnetKostnadsberegnetBidragResultat) =
        BeregnetKostnadsberegnetBidragResultatCore(
            resultatPeriodeListe = mapResultatPeriode(resultat.resultatPeriodeListe),
            avvikListe = mapAvvik(avvikListe)
        )

    private fun mapUnderholdskostnadPeriodeListe(underholdskostnadPeriodeListeCore: List<UnderholdskostnadPeriodeCore>): List<UnderholdskostnadPeriode> {
        val underholdskostnadPeriodeListe = mutableListOf<UnderholdskostnadPeriode>()
        underholdskostnadPeriodeListeCore.forEach {
            underholdskostnadPeriodeListe.add(
                UnderholdskostnadPeriode(
                    referanse = it.referanse,
                    underholdskostnadPeriode = Periode(datoFom = it.periode.datoFom, datoTil = it.periode.datoTil),
                    belop = it.belop
                )
            )
        }
        return underholdskostnadPeriodeListe
    }

    private fun mapBPsAndelUnderholdskostnadPeriodeListe(
        bPsAndelUnderholdskostnadPeriodeListeCore: List<BPsAndelUnderholdskostnadPeriodeCore>
    ): List<BPsAndelUnderholdskostnadPeriode> {
        val bPsAndelUnderholdskostnadPeriodeListe = mutableListOf<BPsAndelUnderholdskostnadPeriode>()
        bPsAndelUnderholdskostnadPeriodeListeCore.forEach {
            bPsAndelUnderholdskostnadPeriodeListe.add(
                BPsAndelUnderholdskostnadPeriode(
                    referanse = it.referanse,
                    bPsAndelUnderholdskostnadPeriode = Periode(datoFom = it.periode.datoFom, datoTil = it.periode.datoTil),
                    andelProsent = it.andelProsent
                )
            )
        }
        return bPsAndelUnderholdskostnadPeriodeListe
    }

    private fun mapSamvaersfradragPeriodeListe(samvaersfradragPeriodeListeCore: List<SamvaersfradragPeriodeCore>): List<SamvaersfradragPeriode> {
        val samvaersfradragPeriodeListe = mutableListOf<SamvaersfradragPeriode>()
        samvaersfradragPeriodeListeCore.forEach {
            SamvaersfradragPeriode(
                referanse = it.referanse,
                samvaersfradragPeriode = Periode(datoFom = it.periode.datoFom, datoTil = it.periode.datoTil),
                belop = it.belop
            )
        }
        return samvaersfradragPeriodeListe
    }

    private fun mapResultatPeriode(resultatPeriodeListe: List<ResultatPeriode>): List<ResultatPeriodeCore> {
        val resultatPeriodeCoreListe = mutableListOf<ResultatPeriodeCore>()
        resultatPeriodeListe.forEach {
            resultatPeriodeCoreListe.add(
                ResultatPeriodeCore(
                    soknadsbarnPersonId = it.soknadsbarnPersonId,
                    periode = PeriodeCore(datoFom = it.periode.datoFom, datoTil = it.periode.datoTil),
                    resultat = ResultatBeregningCore(it.resultat.belop),
                    grunnlagReferanseListe = mapReferanseListe(it)
                )
            )
        }
        return resultatPeriodeCoreListe
    }

    private fun mapReferanseListe(resultatPeriode: ResultatPeriode): List<String> {
        val (underholdskostnad, bPsAndelUnderholdskostnad, samvaersfradrag) = resultatPeriode.grunnlag
        val referanseListe = mutableListOf<String>()
        referanseListe.add(underholdskostnad.referanse)
        referanseListe.add(bPsAndelUnderholdskostnad.referanse)
        referanseListe.add(samvaersfradrag.referanse)
        return referanseListe.sorted()
    }
}
